src/v/serde/avro/parser.h

```diff
@@ -0,0 +1,96 @@
+#pragma once
+
+#include "bytes/iobuf.h"
+#include "container/fragmented_vector.h"
+
+#include <avro/Schema.hh>
+
+#include <variant>
+
+namespace serde::avro {
+
+namespace parsed {
+
+struct record;
+struct list;
+struct map;
+struct avro_null {};
+struct avro_union;
+/**
+ * Avro primitive being either a number, enum, string, bytes, avro fixed or null
+ */
+using primitive = std::variant<
+  avro_null,
+  double,
+  float,
+  int32_t, // Can be an enum value
+  int64_t,
+  bool,
+  iobuf // iobuf can represent either bytes, string or fixed types
+  >;
+
+/**
+ * Variant representing a result of Avro message deserialization
+ */
+using message = std::variant<
+  primitive,
+  std::unique_ptr<map>,
```

Comment:
Why are we using unique_ptr here and in the return type of parse? I thought initially that it has to do with recursive types but that [shouldn't be a problem](https://godbolt.org/z/E5eEr35b6) either?

Reply:
There have been recent fixes to `chunked_vector` to allow forward declared elements. Removing the extra indirection sounds good 👍

Reply:
The only problem is with `union` i am wondering if make it an exception and use unique ptr in union only or keep it consistent as is ?

Reply:
i removed one level of indirection from the message variant

Reply:
I think this is much cleaner thanks Michal (and Nicole for pointing out!).